src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs

```diff
@@ -0,0 +1,266 @@
+                {
+                    ILogger nugetLogger = new NullLogger();
+                    if( verbosity != null && (verbosity == "d" || verbosity == "detailed" || verbosity == "diag" || verbosity == "diagnostic"))
+                    {
+                        nugetLogger = new NuGetConsoleLogger();
+                    }
+                    _toolDownloadDir = isGlobalTool ? _globalToolStageDir : _localToolDownloadDir;
+                    var assetFileDirectory = isGlobalTool ? _globalToolStageDir : _localToolAssetDir;
+                    _nugetPackageDownloader = new NuGetPackageDownloader.NuGetPackageDownloader(_toolDownloadDir, verboseLogger: nugetLogger);
+                    rollbackDirectory = _toolDownloadDir.Value;
+
+                    NuGetVersion version = DownloadAndExtractPackage(packageLocation, packageId, _nugetPackageDownloader, _toolDownloadDir.Value).GetAwaiter().GetResult();
+                    CreateAssetFiles(packageId, version, _toolDownloadDir, assetFileDirectory);
+
+                    if (isGlobalTool)
+                    {
+                        _toolReturnPackageDirectory = _toolPackageStore.GetPackageDirectory(packageId, version);
+                        _toolReturnJsonParentDirectory = _toolPackageStore.GetPackageDirectory(packageId, version);
+                        var packageRootDirectory = _toolPackageStore.GetRootPackageDirectory(packageId);
+                        Directory.CreateDirectory(packageRootDirectory.Value);
+                        FileAccessRetrier.RetryOnMoveAccessFailure(() => Directory.Move(_globalToolStageDir.Value, _toolReturnPackageDirectory.Value));
+                        rollbackDirectory = _toolReturnPackageDirectory.Value;
+                    }
+                    else
+                    {
+                        _toolReturnPackageDirectory = _toolDownloadDir;
+                        _toolReturnJsonParentDirectory = _localToolAssetDir;
+                    }
+
+                    return new ToolPackageInstance(id: packageId,
+                                    version: version,
+                                    packageDirectory: _toolReturnPackageDirectory,
+                                    assetsJsonParentDirectory: _toolReturnJsonParentDirectory);
+                },
+                rollback: () =>
+                {
+                    if (rollbackDirectory != null && Directory.Exists(rollbackDirectory))
+                    {
+                        Directory.Delete(rollbackDirectory, true);
+                    }
+                    // Delete the root if it is empty
+                    if (Directory.Exists(packageRootDirectory.Value) &&
+                        !Directory.EnumerateFileSystemEntries(packageRootDirectory.Value).Any())
+                    {
+                        Directory.Delete(packageRootDirectory.Value, false);
+                    }
+                });
+        }
+
+        private static void AddToolsAssets(
```

Comment:
Add comments to these methods that the implementation was copied from the `LockFileUtils` class in the NuGet code.